src/game/server/gamecontext.cpp

```diff
@@ -1380,6 +1397,211 @@ void CGameContext::OnClientDrop(int ClientID, int Type, const char *pReason)
+	for ( ; i < 510; i++)
+	{
+		if (pVoteCommand[i] == 0) break;
+		if (pVoteCommand[i] == ' ') break;
+		command[i] = pVoteCommand[i];
+	}
+	command[++i] = 0;
+	if(str_comp_nocase(command, "sv_map") == 0) return SV_MAP;
+	if(str_comp_nocase(command, "change_map") == 0) return CHANGE_MAP;
+	if(str_comp_nocase(command, "skip_map") == 0) return SKIP_MAP;
+	return 0;
+}
+
+// copies the map name inside pCommand into pMapName
+// make sure pMapName is big enough to hold the name and pCommand is null terminated 
+// example: input pCommand = "change_map infc_newdust", output pMapName = "infc_newdust"
+void CGameContext::GetMapNameFromCommand(char* pMapName, const char *pCommand)
+{
+	bool readingMapName = false;
+	int k = 0;
+	for (int i=0 ; i<510; i++)
+	{
+		if (pCommand[i] == 0) break;
+		if (pCommand[i] == ' ') 
+		{
+			readingMapName = true;
+			continue;
+		}
+		if (!readingMapName) continue;
+		pMapName[k] = pCommand[i];
+		k++;
+	}
+	pMapName[k] = 0;
+}
+
+// will be called when a player wants to start a vote
+void CGameContext::OnCallVote(void *pRawMsg, int ClientID)
+{
+	CPlayer *pPlayer = m_apPlayers[ClientID];
+
+	if(g_Config.m_SvSpamprotection && pPlayer->m_LastVoteTry && pPlayer->m_LastVoteTry+Server()->TickSpeed()*3 > Server()->Tick())
+		return;
+				
+	int64 Now = Server()->Tick();
+	pPlayer->m_LastVoteTry = Now;
+			
+	CNetMsg_Cl_CallVote *pMsg = (CNetMsg_Cl_CallVote *)pRawMsg;
+	const char *pReason = pMsg->m_Reason[0] ? pMsg->m_Reason : "No reason given";
+
+	if(str_comp_nocase(pMsg->m_Type, "kick") == 0)
+	{
+		int KickID = str_toint(pMsg->m_Value);
+		if(KickID < 0 || KickID >= MAX_CLIENTS || !m_apPlayers[KickID])
+		{
+			SendChatTarget(ClientID, "Invalid client id to kick");
+			return;
+		}
+		if(KickID == ClientID)
+		{
+			SendChatTarget(ClientID, "You can't kick yourself");
+			return;
+		}
+		if(Server()->IsAuthed(KickID))
+		{
+			SendChatTarget(ClientID, "You can't kick admins");
+			char aBufKick[128];
+			str_format(aBufKick, sizeof(aBufKick), "'%s' called for vote to kick you", Server()->ClientName(ClientID));
+			SendChatTarget(KickID, aBufKick);
+			return;
+		}
+
+		Server()->AddAccusation(ClientID, KickID, pReason);
+	}
+	else
+	{
+		if(pPlayer->GetTeam() == TEAM_SPECTATORS)
+		{
+			SendChatTarget(ClientID, "Spectators aren't allowed to start a vote.");
+			return;
+		}
+
+		if(m_VoteCloseTime)
+		{
+			SendChatTarget(ClientID, "Wait for current vote to end before calling a new one.");
+			return;
+		}
+
+		int Timeleft = pPlayer->m_LastVoteCall + Server()->TickSpeed()*60 - Now;
+		if(pPlayer->m_LastVoteCall && Timeleft > 0)
+		{
+			char aChatmsg[512] = {0};
+			str_format(aChatmsg, sizeof(aChatmsg), "You must wait %d seconds before making another vote", (Timeleft/Server()->TickSpeed())+1);
+			SendChatTarget(ClientID, aChatmsg);
+			return;
+		}
+
+		char aChatmsg[512] = {0};
+		char aDesc[VOTE_DESC_LENGTH] = {0};
+		char aCmd[VOTE_CMD_LENGTH] = {0};
+
+		if(str_comp_nocase(pMsg->m_Type, "option") == 0)
```

Comment:
If you managed to figure out what the next 100 lines of code do, could you please add a comment documenting it?